File: src/db/repositories/adjustment_repo.py
import sqlite3
from datetime import datetime, timezone

from src.db.repositories import period_repo
from src.services import audit_service

VALID_TYPES = {"reclassification", "accrual", "provision", "depreciation", "tax", "management",
               "ic_elimination", "opening_correction", "inventory_movement", "other"}


def create_or_replace(conn: sqlite3.Connection, period_id: int, external_ref: str, adjustment_type: str,
                        narration: str, lines: list[dict], user: str | None = None,
                        copied_from_period_id: int | None = None) -> int:
    """Upsert by (period_id, external_ref): if it exists, replace its lines wholesale
    (matches how the Mapping.xlsx round-trip treats re-imports) and log to the audit trail."""
    existing = conn.execute(
        "SELECT adjustment_id FROM adjustments WHERE period_id = ? AND external_ref = ?",
        (period_id, external_ref),
    ).fetchone()

    if existing:
        adjustment_id = existing["adjustment_id"]
        conn.execute("DELETE FROM adjustment_lines WHERE adjustment_id = ?", (adjustment_id,))
        conn.execute(
            """UPDATE adjustments SET adjustment_type = ?, narration = ?, status = 'draft'
               WHERE adjustment_id = ?""",
            (adjustment_type, narration, adjustment_id),
        )
        audit_service.log(conn, "update", "adjustments", str(adjustment_id), new_value=narration, user=user)
    else:
        cur = conn.execute(
            """INSERT INTO adjustments
               (period_id, external_ref, adjustment_type, narration, created_at, created_by,
                copied_from_period_id, status)
               VALUES (?, ?, ?, ?, ?, ?, ?, 'draft')""",
            (period_id, external_ref, adjustment_type, narration,
             datetime.now(timezone.utc).isoformat(), user, copied_from_period_id),
        )
        adjustment_id = cur.lastrowid
        audit_service.log(conn, "create", "adjustments", str(adjustment_id), new_value=narration, user=user)

    for line in lines:
        conn.execute(
            """INSERT INTO adjustment_lines
               (adjustment_id, entity_id, group_account_id, debit_amount, credit_amount, line_narration)
               VALUES (?, ?, ?, ?, ?, ?)""",
            (adjustment_id, line.get("entity_id"), line["group_account_id"],
             line.get("debit_amount", 0.0), line.get("credit_amount", 0.0), line.get("line_narration")),
        )

    return adjustment_id


def get_lines(conn: sqlite3.Connection, adjustment_id: int) -> list[sqlite3.Row]:
    return conn.execute(
        "SELECT * FROM adjustment_lines WHERE adjustment_id = ?", (adjustment_id,)
    ).fetchall()


def list_for_period(conn: sqlite3.Connection, period_id: int) -> list[sqlite3.Row]:
    return conn.execute(
        "SELECT * FROM adjustments WHERE period_id = ? ORDER BY adjustment_id", (period_id,)
    ).fetchall()


def apply_all(conn: sqlite3.Connection, period_id: int, user: str | None = None) -> int:
    period_repo.require_open(conn, period_id, "apply adjustments")
    rows = conn.execute(
        "SELECT adjustment_id FROM adjustments WHERE period_id = ? AND status = 'draft'", (period_id,)
    ).fetchall()
    for r in rows:
        conn.execute("UPDATE adjustments SET status = 'applied' WHERE adjustment_id = ?", (r["adjustment_id"],))
        audit_service.log(conn, "apply", "adjustments", str(r["adjustment_id"]), user=user)
    return len(rows)


def delete(conn: sqlite3.Connection, period_id: int, adjustment_id: int, user: str | None = None) -> None:
    period_repo.require_open(conn, period_id, "delete an adjustment")
    conn.execute("DELETE FROM adjustment_lines WHERE adjustment_id = ?", (adjustment_id,))
    conn.execute("DELETE FROM adjustments WHERE adjustment_id = ?", (adjustment_id,))
    audit_service.log(conn, "delete", "adjustments", str(adjustment_id), user=user)

File: docs/src/db/adjustment-repo.js
// Port of src/db/repositories/adjustment_repo.py
import { persistence } from "./persistence.js";
import * as periodRepo from "./period-repo.js";
import { log as auditLog } from "../services/audit-service.js";

export const VALID_TYPES = new Set([
  "reclassification",
  "accrual",
  "provision",
  "depreciation",
  "tax",
  "management",
  "ic_elimination",
  "opening_correction",
  "inventory_movement",
  "other",
]);

// Upsert by (period_id, external_ref): if it exists, replace its lines wholesale and log
// to the audit trail.
export function createOrReplace(periodId, externalRef, adjustmentType, narration, lines, user = null, copiedFromPeriodId = null) {
  const existing = persistence.get("SELECT adjustment_id FROM adjustments WHERE period_id = ? AND external_ref = ?", [
    periodId,
    externalRef,
  ]);

  let adjustmentId;
  if (existing) {
    adjustmentId = existing.adjustment_id;
    persistence.run("DELETE FROM adjustment_lines WHERE adjustment_id = ?", [adjustmentId]);
    persistence.run(`UPDATE adjustments SET adjustment_type = ?, narration = ?, status = 'draft' WHERE adjustment_id = ?`, [
      adjustmentType,
      narration,
      adjustmentId,
    ]);
    auditLog("update", "adjustments", String(adjustmentId), { newValue: narration, user });
  } else {
    adjustmentId = persistence.insert(
      `INSERT INTO adjustments
         (period_id, external_ref, adjustment_type, narration, created_at, created_by,
          copied_from_period_id, status)
       VALUES (?, ?, ?, ?, ?, ?, ?, 'draft')`,
      [periodId, externalRef, adjustmentType, narration, new Date().toISOString(), user, copiedFromPeriodId]
    );
    auditLog("create", "adjustments", String(adjustmentId), { newValue: narration, user });
  }

  for (const line of lines) {
    persistence.run(
      `INSERT INTO adjustment_lines
         (adjustment_id, entity_id, group_account_id, debit_amount, credit_amount, line_narration)
       VALUES (?, ?, ?, ?, ?, ?)`,
      [
        adjustmentId,
        line.entityId ?? null,
        line.groupAccountId,
        line.debitAmount ?? 0.0,
        line.creditAmount ?? 0.0,
        line.lineNarration ?? null,
      ]
    );
  }

  return adjustmentId;
}

export function getLines(adjustmentId) {
  return persistence.all("SELECT * FROM adjustment_lines WHERE adjustment_id = ?", [adjustmentId]);
}

export function listForPeriod(periodId) {
  return persistence.all("SELECT * FROM adjustments WHERE period_id = ? ORDER BY adjustment_id", [periodId]);
}

export function applyAll(periodId, user = null) {
  periodRepo.requireOpen(periodId, "apply adjustments");
  const rows = persistence.all("SELECT adjustment_id FROM adjustments WHERE period_id = ? AND status = 'draft'", [
    periodId,
  ]);
  for (const r of rows) {
    persistence.run("UPDATE adjustments SET status = 'applied' WHERE adjustment_id = ?", [r.adjustment_id]);
    auditLog("apply", "adjustments", String(r.adjustment_id), { user });
  }
  return rows.length;
}

export function deleteAdjustment(periodId, adjustmentId, user = null) {
  periodRepo.requireOpen(periodId, "delete an adjustment");
  persistence.run("DELETE FROM adjustment_lines WHERE adjustment_id = ?", [adjustmentId]);
  persistence.run("DELETE FROM adjustments WHERE adjustment_id = ?", [adjustmentId]);
  auditLog("delete", "adjustments", String(adjustmentId), { user });
}
